### ISSUE TITLE
Mask RCNN : mAP value of tensorrtx model is lower than the detectron2 python model

### ISSUE THREAD

**Issue by rgkannan676** [wang-xinyu/tensorrtx#943]
Hi @freedenS , Thank you for your work.

- Have you tested the tensorrtx model mAP with coco dataset? I found a difference in the final mAP. 

- I used the python model provided in https://github.com/facebookresearch/detectron2/blob/main/MODEL_ZOO.md ( MaskRcnn [R101-C4] ) . Following is the mAP values.
```

[03/22 17:13:13 d2.evaluation.fast_eval_api]: Evaluate annotation type *bbox*
[03/22 17:13:23 d2.evaluation.fast_eval_api]: COCOeval_opt.evaluate() finished in 9.85 seconds.
[03/22 17:13:23 d2.evaluation.fast_eval_api]: Accumulating evaluation results...
[03/22 17:13:25 d2.evaluation.fast_eval_api]: COCOeval_opt.accumulate() finished in 1.79 seconds.
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.426
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.621
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.460
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.232
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.471
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.576
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.348
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.543
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.568
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.352
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.617
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.739

```

- I converted the same pretrained model to '.wts' file  and to engine file. I tested the model on same coco dataset and got the following mAP.  (SCORE_THRESH=0.05, NMS_THRESH_TEST= 0.5).

```

Evaluate annotation type *bbox*
DONE (t=20.05s).
Accumulating evaluation results...
DONE (t=4.07s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.349
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.506
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.381
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.146
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.380
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.534
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.335
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.435
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.438
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.208
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.471
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.646

```

- You can observe a big difference in the mAP. Is this difference expected?  Can you please advise? Thank you.

**Comment by freedenS** [comment]
I just test about less than 10 images and it shows the diff is less than 1e-5 (fp32 vs fp32)
here are some advice:
1 use detectron2 v0.4
2 make sure you use same opencv (for example v4.2) or you can also save the input data into file (after all preprocess) in detectron2 and use it in trt
3 dump the output of every layer

**Comment by zhuimeng2080** [comment]
On win10 system, RTX 3090, however, for the fp16 engine of mask rcnn model, the cost time of do inference for one frame is about 200ms, why is it so slow?

**Comment by freedenS** [comment]
> On win10 system, RTX 3090, however, for the fp16 engine of mask rcnn model, the cost time of do inference for one frame is about 200ms, why is it so slow?

what's the batch and input size?

**Comment by zhuimeng2080** [comment]
the batch is one, input size is 1280*720.


testing mask_rcnn_R_50_C4_1x.yaml model


| #defineDEVICE0 |
| | #defineBATCH_SIZE1 |
| | #defineBACKBONE_RESNETTYPE R50 |
| | // data |
| | staticconst std::vector<float> PIXEL_MEAN = { 103.53, 116.28, 123.675 }; |
| | staticconst std::vector<float> PIXEL_STD = {1.0, 1.0, 1.0}; |
| | staticconstexprfloat MIN_SIZE = 800.0; |
| | staticconstexprfloat MAX_SIZE = 1333.0; |
| | staticconstexprint NUM_CLASSES = 80; |
| | staticconstexprint INPUT_H = 480; |
| | staticconstexprint INPUT_W = 640; |
| | staticint IMAGE_HEIGHT = 800; |
| | staticint IMAGE_WIDTH = 1333; |
| | // backbone |
| | staticconstint RES2_OUT_CHANNELS = (BACKBONE_RESNETTYPE == R18 || |
| | BACKBONE_RESNETTYPE == R34) ? 64 : 256; |
| | // rpn |
| | staticconst std::vector<float> ANCHOR_SIZES = { 32, 64, 128, 256, 512 }; |
| | staticconst std::vector<float> ASPECT_RATIOS = { 0.5, 1.0, 2.0 }; |
| | staticconstexprint PRE_NMS_TOP_K_TEST = 6000; |
| | staticconstexprfloat RPN_NMS_THRESH = 0.7; |
| | staticconstexprint POST_NMS_TOPK = 1000; |













在 2022-05-30 12:18:55，"JumpPandaer" ***@***.***> 写道：

On win10 system, RTX 3090, however, for the fp16 engine of mask rcnn model, the cost time of do inference for one frame is about 200ms, why is it so slow?

what's the batch and input size?

**Comment by freedenS** [comment]
do you got the latency after the warmup? what's the latency in pytorch?
many reasons may cause this problem, so i can't give a explicit solution.

**Comment by rgkannan676** [comment]
This issue is resolved. It was a mistake in code from my side. Thank you
